### ISSUE TITLE
Problem in read_data of utils.py

### ISSUE THREAD

**Issue by 12baby21** [Atomu2014/pnn-demo#10]
Hi,
  I have downloaded the codes and have successfully gained the ipinyou datasets according to the guide. I chose the fold 2997 as my test datasets. When running main.py, I encountered the following two problems, which are in the function read_data. 
1. 
The origin code is `D_i = [int(x.split(':')[1]) for x in fields[1:]]`, then the problem is that **the list index out of range**. I tried to solve the problem by modifying the code:  `D_i = [int(x.split(':')[1]) for x in fields[2:]]` and changed X_i in the same way. The problem seemed to be solved, but I really don't know whether it is a right operation.
2.
After the previous operation, I encountered the second problem which I am not able to handle. When running this line `X = coo_matrix((coo_data, (coo_rows, coo_cols)), shape=shape)`, python gave me this commention: **ValueError: column index exceeds matrix dimensions**. I really had no means for the problem.

Please give me some guides! Thank you very much!

**Comment by 12baby21** [comment]
The new code for read_data is as follows:
`def read_data(file_name):
    X = []
    D = []
    y = []
    with open(file_name) as fin:
        for line in fin:
            fields = line.strip().split()   #fields是每行元素，去掉“：”后，组成的list
            y_i = int(fields[0])            #列index=0的元素
            X_i = [int(x.split(':')[0]) for x in fields[2:]]  #x组成的list，x是fields这个list的输入list，从index=1到最后
            D_i = [int(x.split(':')[1]) for x in fields[2:]]
            y.append(y_i)
            X.append(X_i)
            D.append(D_i)

    y = np.reshape(np.array(y), [-1])
    #X = libsvm_2_coo(zip(X, D), (len(X), INPUT_DIM)).tocsr()
    X = libsvm_2_coo(zip(X, D), (len(X), INPUT_DIM)).tocsr()
    return X, y`

**Comment by Atomu2014** [comment]
1. I remember I remove the "user_tag" filed in my https://github.com/Atomu2014/make-ipinyou-data repository. I guess in your data format, the fileds[1] contains no ":".

2. I guess the problem comes from feature indexing. You can check the "FIELD_SIZES", "INPUT_DIM", and the maximum value in X. In principle, max(X) should < "INPUT_DIM".

**Comment by 12baby21** [comment]
1. You are absolutely right. In my "train.yzx.txt" or "test.yzx.txt", the fields[1] exists no ":".  But I followed the steps to make ipinyou datasets, and gained the same file names in my folders  just like the example in step 2. I note that, you have slightly changed the data format and feature alignment. So, I guess that I maybe followed the steps at https://github.com/wnzhang/make-ipinyou-data. Is this a possible reason for wrong use of the datasets.
2. I have checked those elements. Actually, it violated the principles. So, does the violation comes from the problem in "1.".
